crates/sui-graphql-rpc/src/context_data/package_cache.rs

```diff
@@ -0,0 +1,1237 @@
+struct StructRef<'m, 'n> {
+    package: AccountAddress,
+    module: Cow<'m, str>,
+    name: Cow<'n, str>,
+}
+
+/// A `StructRef` that owns its strings.
+type StructKey = StructRef<'static, 'static>;
+
+/// Deserialized representation of a type signature that could appear as a field type for a struct.
+/// Signatures refer to structs at their runtime IDs and can contain references to free type
+/// parameters but will not contain reference types.
+#[derive(Clone, Debug)]
+enum OpenSignature {
+    Address,
+    Bool,
+    U8,
+    U16,
+    U32,
+    U64,
+    U128,
+    U256,
+    Vector(Box<OpenSignature>),
+    Struct(StructKey, Vec<OpenSignature>),
+    TypeParameter(u16),
+}
+
+/// Information necessary to convert a type tag into a type layout.
+#[derive(Debug, Default)]
+struct ResolutionContext {
+    /// Definitions (field information) for structs referred to by types added to this context.
+    structs: BTreeMap<StructKey, StructDef>,
+}
+
+/// Interface to abstract over access to a store of live packages.  Used to override the default
+/// store during testing.
+#[async_trait]
+trait PackageStore {
+    /// Latest version of the object at `id`.
+    async fn version(&self, id: AccountAddress) -> Result<SequenceNumber>;
+
+    /// Read package contents.  Fails if `id` is not an object, not a package, or is malformed in
+    /// some way.
+    async fn fetch(&self, id: AccountAddress) -> Result<Package>;
+}
+
+impl PackageCache {
+    fn with_store(store: Box<dyn PackageStore + Send + Sync>) -> Self {
+        let packages = Mutex::new(LruCache::new(PACKAGE_CACHE_SIZE));
+        Self { packages, store }
+    }
+
+    /// Return the type layout corresponding to the given type tag.  The layout always refers to
+    /// structs in terms of their defining ID (i.e. their package ID always points to the first
+    /// package that introduced them).
+    pub(crate) async fn type_layout(&self, mut tag: TypeTag) -> Result<MoveTypeLayout> {
+        let mut context = ResolutionContext::default();
+
+        // (1). Fetch all the information from this cache that is necessary to resolve types
+        // referenced by this tag.
+        context.add_type_tag(&mut tag, self).await?;
+
+        // (2). Use that information to resolve the tag into a layout.
+        context.resolve_type_tag(&tag)
+    }
+
+    /// Return a deserialized representation of the package with ObjectID `id` on-chain.  Attempts
+    /// to fetch this package from the cache, and if that fails, fetches it from the underlying data
+    /// source and updates the cache.
+    async fn package(&self, id: AccountAddress) -> Result<Arc<Package>> {
+        let candidate = {
+            // Release the lock after getting the package
+            let mut packages = self.packages.lock().unwrap();
+            packages.get(&id).map(Arc::clone)
+        };
+
+        // System packages can be invalidated in the cache if a newer version exists.
```

Comment:
The version comes from the system package object, rather than the object whose type we care about, so it changes at most once a day (at the epoch boundary), which is somewhat tolerable (although it is annoying that we need to do this version check -- which is a DB round trip -- each time we reference a system type).

Long-term, we'll put a layout cache in front of this, which doesn't need to be invalidated and should "soak up" most of the requests that would otherwise hit the DB.